### PR TITLE
Translation averaging

### PR DESCRIPTION
**Pull request by akshay-krishnan**
This PR introduces a translation averaging method from [Wilson and Snavely](https://research.cs.cornell.edu/1dsfm/docs/1DSfM_ECCV14.pdf). This method uses the chordal distance to optimize for global translations of a set of cameras from pairwise translation direction measurements. 

Changes introduced: 
- A translation factor that computes the chordal distance between two translation directions.
- A translation recovery class that can build a factor graph from pairwise translation directions and optimize for the global translations using an LM optimizer.

---
This change is [Reviewable](https://reviewable.io/reviews/borglab/gtsam/367)

### CODE REVIEW

**Comment by dellaert** [comment]
Adding different reviewers as we made this code together, and @ProfFan is working furiously on 24 figures for another project. @binitshah and @ayushbaid could you take a hard look?

File: gtsam/sfm/TranslationFactor.h [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+#include <gtsam/geometry/Point3.h>
+#include <gtsam/geometry/Unit3.h>
+#include <gtsam/linear/NoiseModel.h>
+#include <gtsam/nonlinear/NonlinearFactor.h>
+
+namespace gtsam {
+
+/**
+ * Binary factor for a relative translation direction measurement
+ * w_aZb. The measurement equation is
```
Comment (ayushbaid) [inline gtsam/sfm/TranslationFactor.h:30]: Does Z signify 'to' (like a2b)?

Comment (mikesheffler) [reply gtsam/sfm/TranslationFactor.h:30]: Not quite. Check out [a discussion in the Google group](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/gtsam-users/erQkDb2LBf0/m4Pz0lWbFQAJ) from a while back for some more details. It's pretty common, in general, for measurement to be styled `z` and the placement of `b` and `w_a` suggests `b` _to_ `w_a` (and not the other way around) 🙂

Comment (akshay-krishnan) [reply gtsam/sfm/TranslationFactor.h:30]: Yes, Z stands for measurement. w_aZb. In this case, it is the translation direction of b from a, but in the world coordinate frame (aligned with the world's coordinate axes, hence the w_).

**Comment by akshay-krishnan** [comment]
The previous commit should fix the build failure. Please re-review if Travis succeeds.

**Comment by ProfFan** [comment]
@binitshah No, I think you should merge in the current develop as the build error is fixed already.

File: gtsam/sfm/TranslationRecovery.h [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+   * @param scale scale for first relative translation which fixes gauge.
+   * @return Values
+   */
+  Values run(const double scale = 1.0) const;
+
+  /**
+   * @brief Simulate translation direction measurements
+   *
+   * @param poses SE(3) ground truth poses stored as Values
+   * @param edges pairs (a,b) for which a measurement w_aZb will be generated.
```
Comment (ayushbaid) [inline gtsam/sfm/TranslationRecovery.h:104]: return in docstring

File: gtsam/sfm/tests/testTranslationFactor.cpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+static const Key kKey1(2), kKey2(1);
+static const Unit3 kMeasured(1, 0, 0);
+
+/* ************************************************************************* */
+TEST(TranslationFactor, Constructor) {
+  TranslationFactor factor(kKey1, kKey2, kMeasured, model);
+}
+
+/* ************************************************************************* */
+TEST(TranslationFactor, Error) {
```
Comment (ayushbaid) [inline gtsam/sfm/tests/testTranslationFactor.cpp:41]: Test for non-zero error too?

File: gtsam/sfm/TranslationRecovery.cpp [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file TranslationRecovery.h
+ * @author Frank Dellaert
+ * @date March 2020
+ * @brief test recovering translations when rotations are given.
+ */
+
+#include <gtsam/sfm/TranslationRecovery.h>
```
[the remark this answers is not in the document]
Comment (akshay-krishnan) [reply gtsam/sfm/TranslationRecovery.cpp:19]: One of them is TranslationRecovery and the other is TranslationFactor. We separate includes into 3 sections: 
- The header file which the prototype of the class that is being defined here. 
- Headers from other libraries in GTSAM. 
- Headers from standard C++ libraries. 
Each section is alphabetically ordered.

**Comment by ayushbaid** [comment]
LGTM!

**Comment by ProfFan** [comment]
@akshay-krishnan Could you merge in latest develop?

**Comment by ProfFan** [comment]
Note that this breaks `GTSAM_TYPEDEF_POINTS_TO_VECTORS`, by using `xx.normalized(H)` not `normalize(xx, H)`

**Comment by dellaert** [comment]
> Note that this breaks `GTSAM_TYPEDEF_POINTS_TO_VECTORS`, by using `xx.normalized(H)` not `normalize(xx, H)`

Hmmm. Then CI should fail, right?

**Comment by ProfFan** [comment]
Yeah, CI failed on the Pybind wrapper side (we are using that flag) :)

**Pull request merged**